### ISSUE TITLE
Adding tags to append-only REST repo snapshot fails but shows tags in snapshot list

### ISSUE THREAD

**Issue by Sebi2020** [restic/restic#3448]
Output of `restic version`
--------------------------
restic 0.12.0 compiled with go1.15.8 on windows/amd64

How did you run restic exactly?
-------------------------------
`restic tag --add doc [snapshot ids]`

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
REST-Server on Linux Ubuntu 18 in append-only mode behind a https reverse-proxy.

Expected behavior
-----------------

Adding tags should fail completly or should be successful

Actual behavior
---------------

The client fails to add the tags and is retrying to rewrite it in an infinite loop with the following messages:

```
repository d0beffd4 opened successfully, password is correct
create exclusive lock for repository
Remove(<snapshot/9dd62f68ce>) returned error, retrying after 552.330144ms: blob not removed, server response: 403 Forbidden (403)
Remove(<snapshot/9dd62f68ce>) returned error, retrying after 1.080381816s: blob not removed, server response: 403 Forbidden (403)
signal interrupt received, cleaning up
unable to modify the tags for snapshot ID "9dd62f68ce7be6e634f3f79514b39832c21d23ddf1112ae9824692c3a5e5a3df", ignoring: context canceled
no snapshots were modified
```

But if you open the snapshot list with `restic snapshots` afterwards, it looks like restic has added the tags:

```
b8abcd8f  2021-06-28 15:09:02  My-PC     doc        /path1
                                                                                  /path2
```

Steps to reproduce the behavior
-------------------------------
1.  Deploy an append-only rest server with private_repo enabled
2. Set `RESTIC_REPOSITORY` to your rest server repo
3. Create a backup with `restic backup path1 path2 ...`
4. Try to add tags to the created snapshot afterwards with `restic tag --add mytag snapshot_id`
5. Cancel process
6. Run `restic snapshots` the list contians the new tag, although restic previously said it failed  

Do you have any idea what may have caused this?
-----------------------------------------------
I have no clue, but maybe some caching causes restic snapshots to display the tags. I've tried to run restic with the --cleanup-cache flag, but this hadn't changed anything.

Do you have an idea how to solve the issue?
-------------------------------------------
No

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes

**Comment by rawtaz** [comment]
If you look at your snapshots list before and after the tag adding command, you will probably see that a new snapshot with the tag was added, and the snapshot you referenced is still there after running the command.

Since restic doesn't *modify* the snapshots, what it does in practice is that it first adds the tag to a new snapshot (which is basically a copy of the one you reference), and once that is done tries to delete the "old" snapshot you referenced.

Since `--append-only` mode lets restic *add* new data/files but doesn't let it *remove* data/files, the result you are seeing is expected. It might be a bit unintuitive, but at the same time if the user thinks in terms of "modifying" the snapshot by adding tags to it, it should already be clear to them that this will not be possible when they run the rest-server with `--append-only` mode, since this mode only let's you *add* things, not *modify* things.

Simply put; If you want to *modify* your repository in any way, you cannot do that when using the `--append-only` option (unless you e.g. do it on the rest-server's filesystem itself that is). For this reason I'm closing this issue now. Thanks for looking into this though!

**Comment by Sebi2020** [comment]
I think if restic recognizes that the server runs in append-only mode, it should stop printing errors and retrying the operation and instead only lets the user now that it cannot remove old data. After running this command the check command also sayed that there are packs files which are not listed by index files.

**Comment by rawtaz** [comment]
It's not that straight forward. It gets a 403 error. Should it treat that as "I will not have permissions for the foreseeable future" or "right now I don't have permission, I will try again in a bit in case it works then"? It doesn't know of rest-server's `--append-only` mode, it's backend agnostic.

**Comment by Sebi2020** [comment]
It looks like it does not work like expected:

>  After running this command `restic tag -add ...` the check command also said that there are packs files which are not listed by index files. I had to run prune on the server which removed a bunch of files.

How can this happen if the REST-Server does not modify files and is just adding new files.

**Comment by rawtaz** [comment]
Please show that output. It's probably not a problem, you'd get that with e.g. a backup run that's cancelled prematurely.

**Comment by Sebi2020** [comment]
@rawtaz 

> [...] it should already be clear to them that this will not be possible when they run the rest-server with --append-only mode, since this mode only let's you add things, not modify things.

You imply that the user is always equal to the operator of the server. That don't have to be the case! Maybe the user does not know which servers run in append-only mode and which do not and that restic is backend agnostic does not prevent restic from have special handlings (in the backend code, not the frontend) for special cases. It's just a question of responsibility. The backend knows it's handling a rest server and therefore could check if the server runs in append-only mode (with some modifcations to the rest protocol). If that is the case it can refuse to create a new snapshot with the new tags because deleting the old one will fail. 
From the developer perspective it makes sense to argument like this, but it's not really a user friendly behaviour.

> Please show that output. It's probably not a problem, you'd get that with e.g. a backup run that's cancelled prematurely (restic uploads pack files, and if there backup process is cancelled before a corresponding index file that references these packs is uploaded as well, you will end up with unreferenced packs).

if it happens again, i will show the output. I didn't saved it before pruning the repo.
